### PR TITLE
Refactor BinaryOp symbol lookup; expand operator set and Pratt parser

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Refactors BinaryOp symbol resolution to use a static SYMBOLS lookup table and expands the operator surface as part of the Pratt parser enhancements. Added T_CARET token (^), new unary ops BitNot and Ref, and expanded BinaryOp with Concat, Shl, Shr, BitAnd, BitXor, BitOr, Lt, Lte, Gt, Gte. No behavioral changes beyond syntax/precedence.
- BinaryOp::symbol now reads from a static SYMBOLS lookup table (SYMBOLS). Tests exercise symbol formatting via a dedicated test.

### Core language and tokenizer
- Added T_CARET token (caret) to support the ^ operator.
- Updated tokenizer to recognize caret and map it to SyntaxKind::T_CARET.
- Extended language syntax kinds to include T_CARET.

### AST and operators
- Unary operators now include BitNot and Ref in addition to existing Not and Neg.
- Binary operators extended with: Concat, Shl, Shr, BitAnd, BitXor, BitOr, Lt, Lte, Gt, Gte.
- String representations updated for new unary/binary operators.
- Refactor: BinaryOp::symbol now reads from a static SYMBOLS lookup table (SYMBOLS) instead of a hard-coded match.

### Parser precedence
- Introduced new binding powers for the updated operator set:
  - BP_PREFIX: 80
  - BP_SHIFT: 55
  - BP_BIT_AND: 45
  - BP_BIT_XOR: 40
  - BP_BIT_OR: 35
  - BP_CMP: 30
  - BP_ADD_SUB and related: maintained for existing operators
- Updated INFIX_TABLE to include:
  - T_PLUSPLUS -> Concat (l_bp: BP_ADD_SUB, r_bp: BP_ADD_SUB+1)
  - T_SHL -> Shl (l_bp: BP_SHIFT, r_bp: BP_SHIFT+1)
  - T_SHR -> Shr (l_bp: BP_SHIFT, r_bp: BP_SHIFT+1)
  - T_AMP -> BitAnd (l_bp: BP_BIT_AND, r_bp: BP_BIT_AND+1)
  - T_CARET -> BitXor (l_bp: BP_BIT_XOR, r_bp: BP_BIT_XOR+1)
  - T_PIPE -> BitOr (l_bp: BP_BIT_OR, r_bp: BP_BIT_OR+1)
  - T_LT/T_LTE/T_GT/T_GTE -> Lt/Lte/Gt/Gte (l_bp: BP_CMP, r_bp: BP_CMP+1)

### Tests
- Expanded parser tests to cover new operators and precedence:
  - Concatenation (a ++ b, a + b ++ c, etc.)
  - Shifts (a << b, a >> b, mixed with additions)
  - Bitwise operators (a & b, a ^ b, a | b, and combinations)
  - Comparisons (a < b, a <= b, a > b, a >= b) with proper binding
  - Prefix BitNot (~) and Ref (&) tests
  - Prefix/postfix and implication (=>) interactions with new operators
- Added new test file: src/parser/ast/tests/binary_op_symbols.rs to verify BinaryOp::symbol formatting.
- Updated tests in src/parser/tests/operator_precedence.rs to reflect new precedence rules and operators.

### Documentation
- docs/pratt-parser-for-ddlog-expressions.md updated with: 5.5 Operator table completion
- docs/roadmap.md updated to reflect operator-table completion is done.

### Roadmap
- - [x] Complete the operator table to match the specification, including ++, ^, and =>, with precedence tests.

## Test plan
- [x] Run cargo test to verify parser, AST, and tokenization updates.
- [x] Validate parsing of new operator expressions across precedence and associativity cases.
- [x] Verify tokenizer emits T_CARET for ^ and mapping to T_CARET in parser.

## Documentation and compatibility notes
- The & token now serves dual purposes: as a prefix operator (Ref) and as an infix operator it represents bitwise AND.
- The Pratt parser is extended with separate prefix and infix tables to accommodate operator overloading without breaking existing syntax.

## How to test locally
- Build and run tests:
- cargo test
- Add/modify sample expressions in tests under src/parser/tests/expression.rs to exercise the new operators and their precedence.

📎 Task: https://www.terragonlabs.com/task/ba4176d5-2ed7-4354-ac08-9e2e9be22d82

📎 Task: https://www.terragonlabs.com/task/d6ab252b-8ebd-42a5-8598-ce074c2ab4d9

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[src/parser/ast/expr.rs](https://github.com/leynos/ddlint/pull/192/files/3ed733eb34c7f46b94d66302dbf4ce77299cbee9#diff-f7d3e82ed7f9b01b119c147a71f4e53b7766ff4c4ab7187482ccc11ed861f9d1)

Comment on lines +159 to +160

~~~rust
    fn misc_symbol(self) -> &'static str {
        match self {
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/69278/delta?repo-id=1160463&review-id=192&biomarker=Code+Duplication&filename=src%2Fparser%2Fast%2Fexpr.rs&method=)
The module contains 4 functions with similar structure: BinaryOp.arithmetic_symbol,BinaryOp.bitwise_symbol,BinaryOp.comparison_symbol,BinaryOp.misc_symbol

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Partially Resolved ⚠️

Comments 5, 6, 7 (Testing requirements): PARTIALLY RESOLVED

    Extensive behavioral tests exist in src/parser/tests/expression.rs covering all new operators, precedence, and interactions
    However, these comments specifically requested unit tests for:
        Token→SyntaxKind→UnaryOp/BinaryOp mapping validation (Comment 5)
        Direct binding power assignment verification (Comment 6)
        Tokenizer output validation for ^ and & (Comment 7)
    The behavioral integration tests are comprehensive, but dedicated unit tests for these internal mappings and tokenization stages are not evident

Not Resolved ❌

Comment 3 (Dual-use & edge cases): NOT RESOLVED

    Missing tests for: &a & b, a & &b, &a&b
    These would exercise the tokenizer/Pratt parser's handling of prefix vs infix & in ambiguous contexts

Comment 4 (Comparison operators with ==/!=): NOT RESOLVED

    Missing tests like: a == b < c, a < b == c
    These would verify precedence boundary between new comparison operators and existing equality operators

Comment 8 (Refactor BinaryOp::symbol() for cyclomatic complexity): NOT RESOLVED

    The method currently has 24 match arms, exceeding the limit of 12
    No refactoring has been applied to reduce complexity (e.g., splitting by category or using a data-driven approach)

**Comment by leynos** [comment]
@coderabbitai please create a new GitHub issue for the following:

src/parser/tests/expression.rs (1)

    1-485: Stop growing this test file; split it to satisfy the 400-line limit.

    Move operator-precedence cases (including the newly added block) into a dedicated test module/file and keep this file focused on core expression forms and error reporting. This breach will keep recurring as more operators land.

File: src/parser/ast/expr.rs [hunk 1]
```diff
@@ -76,26 +102,45 @@ pub enum BinaryOp {
...
-            Self::Assign => "=",
-            Self::Seq => ";",
-            Self::Imply => "=>",
-        }
+        Self::SYMBOLS[self as usize]
     }
 }
 
+const _: [(); BinaryOp::SYMBOLS.len()] = [(); (BinaryOp::Imply as usize) + 1];
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/parser/ast/expr.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.
